Cmake: Add MBED_TEST_MODE macro

harmut01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes #14494

MBED_TEST_MODE was used in CLI1 to prevent multiple applications from
being built. Add the macro to provide backward compatibility, along with a test 
application.

#### Impact of changes <!-- Optional -->
None

#### Migration actions required <!-- Optional -->
None

### Documentation <!-- Required -->

CLI1 Reference: https://os.mbed.com/docs/mbed-os/v6.9/debug-test/greentea-for-testing-applications.html

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

harrisonmutai-arm committed:
Cmake: Add backward compatibilty support for MBED_TEST_MODE

Fixes #14494

MBED_TEST_MODE is required for backward compatibilty with CLI1. This
adds a test to ensure that the macro is created when using CLI2 for
testing. This also adds a test in `.travis.yml` that checks whether
CMake defines the macro when BUILD_TESTING is on.

CLI1 Reference: https://os.mbed.com/docs/mbed-os/v6.9/debug-test/greentea-for-testing-applications.html

Also, explicitly create and set the macro BUILD_TESTING to allow for
MBED_TEST_MODE to be defined by CMake. MBED_TEST_MODE is required for
backward compatibilty with CLI1.  BUILD_TESTING is used to determine
whether to define MBED_TEST_MODE.  Normally, this would be automatically
done by CTest (which we intend to add for test automation) but this
hasn't yet been added to our Greentea tests.

Patater commented on .travis.yml:
```diff
@@ -342,4 +355,38 @@ matrix:
+      python: 3.8
+      install:
+        # Hide Travis-preinstalled CMake
+        - sudo rm -rf /usr/local/cmake*
```
Do we need this second hiding here or was the first "Extended" hiding OK?

harmut01 replied:
The "Extended" hiding didn't work the first time because I hadn't added sources to the apt config, it should work now though.

harmut01 replied:
@0xc0170 does it make sense to create a new "CMake" job in our config?

0xc0170 replied:
yes

Patater commented on CMakeLists.txt:
```diff
@@ -72,6 +72,14 @@ target_compile_definitions(mbed-core
         ${MBED_CONFIG_DEFINITIONS}
 )
 
+# Add MBED_TEST_MODE for backward compatibility with Greentea testing in CLI1
```
```diff
- in CLI1
+ with Mbed CLI 1
```

Patater commented on tools/cmake/tests/mbed_test_mode/main.cpp:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if BUILD_TESTING && !defined(MBED_TEST_MODE)
```
Should we put this test inside a `TARGET_IGNORE` folder to make Mbed CLI 1 ignore it? I'm not clear on how we avoid multiple definition of `main` errors without some way of excluding this test from normal builds of applications.

Maybe a path something like this?
```
tools/cmake/tests/TARGET_IGNORE/mbed_test_mode/main.cpp
```

0xc0170 replied:
Isn't tools folder always ignored? If no, we will need unfortunately hide this. although `TESTS` should work as it does for our greentea tests?

Patater replied:
That might be the way this is excluded. If so, nothing to do here. I'd expect CI for examples to break if the hiding weren't working.

LDong-Arm replied:
It's ignored by [tools/.mbedignore](https://github.com/ARMmbed/mbed-os/blob/master/tools/.mbedignore)

Patater commented on CMakeLists.txt:
```diff
@@ -72,6 +72,14 @@ target_compile_definitions(mbed-core
         ${MBED_CONFIG_DEFINITIONS}
 )
 
+# Add MBED_TEST_MODE for Greentea testing backward compatibility with CLI1
```
```diff
- Add MBED_TEST_MODE for Greentea testing backward compatibility with CLI1
+ Add MBED_TEST_MODE for backward compatibility with Greentea tests written for use with Mbed CLI 1
```

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -72,6 +72,14 @@ target_compile_definitions(mbed-core
 )
 
+# Add MBED_TEST_MODE for Greentea testing backward compatibility with CLI1
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME AND BUILD_TESTING)
```
`BUILD_TESTING` is set by CTest, which is used by unit tests only. If I understand correctly, Greentea tests don't use CTest - they are normal Mbed applications with [`greentea-client`](https://github.com/ARMmbed/mbed-os/tree/master/features/frameworks/greentea-client) used as a library, to allow synchronisation/reporting between host and target.

See this Greentea test CMake definition for example: https://github.com/ARMmbed/mbed-os/blob/mbed-os-6.9.0/platform/tests/TESTS/mbed_platform/wait_ns/CMakeLists.txt. CTest is not used.

So I'm not sure this would work for a real Greentea tests.

0xc0170 replied:
Greentea tests require more work. They will use CTest.

As they are now, we should explicitly enable `BUILD_TESTING`, adding into the function `mbed_greentea_add_test()` ?

LDong-Arm replied:
If it'll be used for all test suites, then it makes sense to add it to `mbed_greentea_add_test()`.

LDong-Arm commented on tools/cmake/tests/mbed_test_mode/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+project(mbed-test-mode-check)
+
+set(PROJECT_NAME ${CMAKE_PROJECT_NAME} CACHE INTERNAL "")
+include(CTest)
```
I mean, for a typical/existing Greentea test, we are not including CTest already because we currently don't need any CTest features. Unless we use that for automation in the future?

0xc0170 replied:
I believe we will use that for automation.

LDong-Arm replied:
Cool, thanks

LDong-Arm commented on .travis.yml:
```diff
@@ -343,3 +348,44 @@ matrix:
+        - cmake --build ${ROOT}/cmake_build/${TARGET_NAME}/${PROFILE}/${TOOLCHAIN}/
+    - <<: *cmake-vm
+      name: "Backward compatiblity check - MBED_TEST_MODE"
+      env: NAME=mbed-test-mode-check ROOT=tools/cmake/tests/mbed_test_mode/ TOOLCHAIN=GCC_ARM TARGET_NAME=K64F PROFILE=develop
```
You can move the `env` to line 355, and remove  lines 389-391. The reason is lines 352-388 must be a complete test with all variable defined. And 389-391 is a second test that partially overrides the first. Since we only needs one test, the override isn't needed.

LDong-Arm commented on .travis.yml:
```diff
@@ -35,8 +35,13 @@ before_install:
+    sources:
+      - sourceline: 'deb https://apt.kitware.com/ubuntu/ xenial main'
+        key_url: 'https://apt.kitware.com/keys/kitware-archive-latest.asc'
+      - sourceline: 'deb https://apt.kitware.com/ubuntu/ xenial-rc main'
```
```suggestion
```
We don't need RC anymore. This was mainly used in the early days (last year) when our contributions to the CMake project weren't released yet.

rajkan01 replied:
Just to add, we need to remove this `- sourceline: 'deb https://apt.kitware.com/ubuntu/ xenial-rc main`' from all the examples .travis.yml file as well and to use `focal` dist

LDong-Arm commented on tools/cmake/tests/mbed_test_mode/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+
+add_subdirectory(
+    ${MBED_PATH} build
+)
```
Where is this "build" directory?

Patater replied:
Do we need to `add_subdirectory()` at all?

Patater commented on tools/cmake/tests/mbed_test_mode/main.cpp:
```diff
@@ -0,0 +1,13 @@
+#error "MBED_TEST_MODE not defined with BUILD_TESTING on"
+#else
+int main(){
+    return 1;
```
Should this return 0? Usually 1 means some sort of error, doesn't it?

Patater commented on tools/cmake/tests/mbed_test_mode/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+target_sources(mbed-test-mode-check
+    PRIVATE
+        main.cpp
+)
```
Should we do `add_test(mbed-test-mode-check)` so that this test will get built and run with CTest?

harmut01 replied:
I've added it in the latest commit 👌🏿

harrisonmutai-arm committed:
CMake: Refactor MBED_TEST_MODE test application

LDong-Arm reviewed (approved):
LGTM

harrisonmutai-arm committed:
Remove redundant ninja-build from apt packages

harrisonmutai-arm committed:
Travis: Add kitware repository to source list

The default distribution of CMake that comes with focal is 3.16.3. Add
kitware's repository (https://apt.kitware.com/) to the source list so
the latest copy of CMake is installed during CI.
